2010-10-11  Michael Tautschnig  <mt@debian.org>

	* setup-storage, setup-storage/Parser.pm: Perform additional check for invalid
		configs (such as empty disk_config X stanzas).
Index: trunk/bin/setup-storage
===================================================================
--- trunk.orig/bin/setup-storage
+++ trunk/bin/setup-storage	
@@ -127,6 +127,9 @@
 print "Using config file: $opt_f\n";
 &FAI::run_parser($config_file);
 
+# make sure there are no empty disk_config stanzas
+&FAI::check_config;
+
 # read the sizes and partition tables of all disks listed in $FAI::disks
 &FAI::get_current_disks;
 
Index: trunk/lib/setup-storage/Parser.pm
===================================================================
--- trunk.orig/lib/setup-storage/Parser.pm
+++ trunk/lib/setup-storage/Parser.pm	
@@ -1076,5 +1076,33 @@
   defined $FAI::Parser->file($input) or die "Syntax error\n";
 }
 
+################################################################################
+#
+# @brief Check for invalid configs (despite correct syntax)
+#
+################################################################################
+sub check_config {
+
+  # loop through all configs
+  foreach my $config (keys %FAI::configs) {
+    if ($config =~ /^PHY_(.+)$/) {
+      (scalar(keys %{ $FAI::configs{$config}{partitions} }) > 0) or
+        die "Empty disk_config stanza for device $1\n";
+    } elsif ($config =~ /^VG_(.+)$/) {
+      next if ($1 eq "--ANY--");
+      next;
+    } elsif ($config eq "RAID") {
+      (scalar(keys %{ $FAI::configs{$config}{volumes} }) > 0) or
+        die "Empty RAID configuration\n";
+    } elsif ($config eq "CRYPT") {
+      next;
+    } elsif ($config eq "TMPFS") {
+      next;
+    } else {
+      &FAI::internal_error("Unexpected key $config");
+    }
+  }
+}
+
 1;
 
